### ISSUE TITLE
takeUntil

### ISSUE THREAD

**Issue by vjpr** [caolan/highland#187]
Is there a way to perform `takeUntil`?

I need to split a file into lines and find the first line that matches a regex. Maybe there is another way to do this?

Cheers.

**Comment by vjpr** [comment]
I think `find` will work better for my example. But if I did want to take until a predicate, is this possible?

**Comment by LewisJEllis** [comment]
I generally see what you're getting at, but can you give an example to clarify the difference between what `find` can do and what you're asking about?

**Comment by vjpr** [comment]
Example: Read first paragraph from a `README.md` file after the heading.

```
s = require 'string'
fs = require 'fs'
h = require 'highland'

filePath = ...
file = h fs.createReadStream filePath, 'utf8'
file.split()
.skipWhile (x) -> s(x).startsWith('#') or s(x).isEmpty()
.takeUntil (x) -> s(x).isEmpty() or s(x).startsWith '#'
```

**Comment by jeromew** [comment]
we seem to have a hard time undestanding each other's on this one (now with coffeescript added and skipWhile)

From my undestanding, in your example @LewisJEllis , `takeUntil` would mean "let the tokens pass through until we hit a predicate", so I understand that your

    .resume(); // 3, 4, 3, 2, 1

should read

    .resume(); // 1, 2

Reading @vjpr's example it seems that the request is to have a sort of state machine to detect the first interesting token (`skipWhile`) and detect the last interesting token (`takeUntil`)

I am not sure we have a simple way to do this as it stands.

This might be an interesting use case for `splitBy` with a function https://github.com/caolan/highland/issues/163.

**Comment by LewisJEllis** [comment]
@jeromew yes, I got that flipped around; my example should output 1, 2. The behavior in my example should probably be named `skipUntil`.

Right now it's possible to just use a closure to build that state machine for filter, something like:
```javascript
function until(pred) {
  var passing = true;
  return function (x) {
    return passing && (passing = !pred(x));
  }
}

function gt2(x) {
  return x > 2;
}

_([1, 2, 3, 4, 3, 2, 1])
  .filter(until(gt2))
  .doto(console.log)
  .resume(); // 1, 2
```
but a slightly more direct implementation is certainly possible, and this could be a common enough pattern to add some support for (these are certainly functions I've used in other languages before). 

Re: naming - among all the combinations of take, skip, while, and until, I think `takeWhile` and `skipWhile` are the two I've seen the most. Thoughts on until vs while?

**Comment by valeriangalliat** [comment]
I was also in the need of (lazy) `takeWhile` and `taleUntil` functions, that is, not keeping to read from the input stream when the stop conditions is met.

Here's my implementation:

```js
_().constructor.prototype.takeWhile = function (f) {
  return this.consume(function (err, x, push, next) {
    if (err) {
      push(err);
      next();
    } else {
      if (f(x)) {
        push(null, x);
        next();
      } else {
        push(null, _.nil);
      }
    }
  });
};

_().constructor.prototype.takeUntil = function (f) {
  return this.takeWhile(function (x) {
    return !f(x);
  });
};
```

And an example usage:

```js
function infiniteRange(start, increment) {
  return _(function (push, next) {
    push(undefined, start);
    start += increment;
    next();
  });
}

infiniteRange(0, 1)
  .takeWhile(function (x) {
    return x < 5;
  })
  .each(console.log);

// 0 1 2 3 4

infiniteRange(0, 1)
  .takeUntil(function (x) {
    return x >= 5;
  })
  .each(console.log);

// 0 1 2 3 4
```

I'm thinking about making a PR with these, do you think such functions would be welcome in Highland?

**Comment by valeriangalliat** [comment]
Since transducers, it's now possible to do this; with my previous example:

```js
var _ = require('highland');
var t = require('transducers-js');

function infiniteRange(start, increment) {
  return _(function (push, next) {
    push(undefined, start);
    start += increment;
    next();
  });
}

infiniteRange(0, 1)
  .transduce(t.takeWhile(function (x) {
    return x < 5;
  }))
  .each(console.log);
// 0 1 2 3 4
```

No `takeUntil` but it was just sugar around `takeWhile`.
